src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2144,6 +2144,38 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
 {
   g_assert (self->loaded);
 
+  /* It dramatically simplifies a lot of the logic below if we
+   * drop the staged deployment from both the source deployment list,
+   * as well as the target list.  We don't want to write it to the bootloader
+   * now, which is mostly what this function is concerned with.
+   * In the future we though should probably adapt things to keep it.
+   */
+  if (self->staged_deployment)
+    {
+      if (!glnx_unlinkat (AT_FDCWD, _OSTREE_SYSROOT_RUNSTATE_STAGED, 0, error))
+        return FALSE;
+
+      if (!_ostree_sysroot_rmrf_deployment (self, self->staged_deployment, cancellable, error))
```

Comment:
So then here, in the non-finalize path, let's say... `ostree admin undeploy 2`, don't we need to keep the staged deployment? I.e. only remove it from the current deployments, but keep it on disk so it gets added back at the final reload?

Reply:
The problem with keeping it is we need to do juggling of the deployment arrays; that's what the comment above is about:

```
   /* It dramatically simplifies a lot of the logic below if we
   * drop the staged deployment from both the source deployment list,
   * as well as the target list.  We don't want to write it to the bootloader
   * now, which is mostly what this function is concerned with.
   * In the future we though should probably adapt things to keep it.
   */
```

Reply:
Right I mean, still remove it from the `self->deployments` array for that purpose, but don't actually remove it on disk. Otherwise, IIUC, even something like `undeploy 2` (let's say we're booted into `1`) will delete the staged deployment, right? That seems a bit surprising from a UX standpoint. Anyway, I'm cool with tweaking this in follow-ups!